### PR TITLE
feat: Introducing Tracing with OpenTelemetry API

### PR DESCRIPTION
**Pull request by jimit-j-shah**
This change introduces Client-Side Tracing using OpenTelemetry API in Datastore Java SDK.

User Guide & Documentation: 

- https://cloud.google.com/datastore/docs/client-side-traces
- https://cloud.google.com/datastore/docs/trace-span-references

Fixes googleapis/java-datastore#1428 ☕️

### CODE REVIEW

**Comment by cindy-peng** [comment]
Hi @jimit-j-shah, is this pull request combining all previous changes made on the feature branch`[tracing-main-3](https://github.com/googleapis/java-datastore/tree/tracing-main-3)`?

**Comment by jimit-j-shah** [comment]
> Hi @jimit-j-shah, is this pull request combining all previous changes made on the feature branch [tracing-main-3](https://github.com/googleapis/java-datastore/tree/tracing-main-3)?

Yes. Just rebased on `main`'s new changes into a new branch.

**Comment by jimit-j-shah** [comment]
Will wait to merge until we have launch approval.

File: .github/workflows/hermetic_library_generation.yaml [hunk 1]
```diff
@@ -17,10 +17,14 @@ name: Hermetic library generation upon generation config change through pull req
...
   pull_request:
 
+env:
+  HEAD_REF: ${{ github.head_ref }}
+  REPO_FULL_NAME: ${{ github.event.pull_request.head.repo.full_name }}
+
 jobs:
   library_generation:
     # skip pull requests coming from a forked repository
-    if: github.event.pull_request.head.repo.full_name == github.repository
```
Comment (cindy-peng) [inline .github/workflows/hermetic_library_generation.yaml:23]: Could we separate these unrelated commits?

File: google-cloud-datastore/pom.xml [hunk 2]
```diff
@@ -160,6 +194,56 @@
...
+    <dependency>
+      <groupId>io.opentelemetry</groupId>
+      <artifactId>opentelemetry-semconv</artifactId>
+      <version>1.1.0-alpha</version>
+      <scope>test</scope>
+    </dependency>
+    <!-- END OpenTelemetry -->
+    <!-- Cloud Ops -->
+    <dependency>
+      <groupId>com.google.cloud.opentelemetry</groupId>
```
Comment (cindy-peng) [inline google-cloud-datastore/pom.xml:231]: Nit: Move this before `<!-- End OpenTelemetry -->`

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 3]
```diff
@@ -125,20 +142,57 @@ void setPrevTransactionId(ByteString transactionId) {
...
+                          io.opentelemetry.api.trace.Span.wrap(
+                              parentSpanContext.getSpanContext())));
+      return otelTraceUtil.addSettingsAttributesToCurrentSpan(spanBuilder).startSpan();
+    }
+
     @Override
     public T call() throws DatastoreException {
-      transaction = datastore.newTransaction(options);
-      try {
+      // TODO Instead of using OTel Spans directly, TraceUtil.Span should be used here. However,
```
Comment (cindy-peng) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:165]: Do we have a github issue tracking this?

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 4]
```diff
@@ -602,20 +692,37 @@ private com.google.datastore.v1.CommitResponse commitMutation(
...
+              getOptions().getClock());
+      span.addEvent(
+          spanName + " complete.",
+          new ImmutableMap.Builder<String, Object>()
+              .put("doc_count", response.getMutationResultsCount())
+              .put("transactional", isTransactional)
+              .put(
+                  "transaction_id",
+                  isTransactional ? requestPb.getTransaction().toStringUtf8() : "")
+              .build());
```
Comment (ehsannas) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:719]: nit: It'd be better to have all attribute keys defined as constants somewhere (similar to [this](https://github.com/googleapis/java-firestore/blob/main/google-cloud-firestore/src/main/java/com/google/cloud/firestore/telemetry/TraceUtil.java#L58-L65))

**Review by ehsannas — COMMENTED** [review]
The public APIs look good to me. I also briefly reviewed certain parts of the PR (skipped the tests). I've pointed out some nits below. I defer the approval to the team that maintains this repo.

**Comment by jimit-j-shah** [comment]
> Hi @jimit-j-shah, is this pull request combining all previous changes made on the feature branch [tracing-main-3](https://github.com/googleapis/java-datastore/tree/tracing-main-3)?

Yes. Now I've done a fresh merge from main in https://github.com/googleapis/java-datastore/pull/1576. Let's continue the conversation there.

**Comment by jimit-j-shah** [comment]
> Will wait to merge until we have launch approval.

We now have launch approval.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java (hunk 3 above)
Comment (jimit-j-shah) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:165]: #1491

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/spi/v1/HttpDatastoreRpc.java [hunk 5]
```diff
@@ -40,6 +39,7 @@
 import com.google.datastore.v1.RunAggregationQueryResponse;
 import com.google.datastore.v1.RunQueryRequest;
 import com.google.datastore.v1.RunQueryResponse;
+import io.opencensus.trace.Tracing;
```
[the remark this answers is not in the document]
Comment (jimit-j-shah) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/spi/v1/HttpDatastoreRpc.java:42]: Replaced `TraceUtil` usage https://github.com/googleapis/java-datastore/commit/45da20b4a502c2bbe5b03d52f843ce5e39174c76

File: google-cloud-datastore/pom.xml [hunk 6]
```diff
@@ -111,14 +128,31 @@
       <artifactId>jsr305</artifactId>
     </dependency>
 
+    <!-- OpenTelemetry -->
+    <dependency>
+      <groupId>io.opentelemetry</groupId>
+      <artifactId>opentelemetry-sdk</artifactId>
```
[the remark this answers is not in the document]
Comment (jimit-j-shah) [reply google-cloud-datastore/pom.xml:134]: Typo - it should have been under `Test Dependencies` - good catch!

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java (hunk 4 above)
Comment (jimit-j-shah) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:719]: Done.

File: google-cloud-datastore/pom.xml (hunk 2 above)
Comment (jimit-j-shah) [reply google-cloud-datastore/pom.xml:231]: This is a cloud dependency, not an opentelemetry dependency.

File: .github/workflows/hermetic_library_generation.yaml (hunk 1 above)
Comment (jimit-j-shah) [reply .github/workflows/hermetic_library_generation.yaml:23]: I never made these commits. Seem to be rebase/merge artifacts?

**Comment by jimit-j-shah** [comment]
This PR is superseded by https://github.com/googleapis/java-datastore/pull/1576 - will implement all the comments there.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 7]
```diff
@@ -125,20 +142,57 @@ void setPrevTransactionId(ByteString transactionId) {
...
     @Override
     public T call() throws DatastoreException {
-      transaction = datastore.newTransaction(options);
-      try {
+      // TODO Instead of using OTel Spans directly, TraceUtil.Span should be used here. However,
+      // the same code in startSpanInternal doesn't work when EnabledTraceUtil.StartSpan is called
+      // probably because of some thread-local caching that is getting lost. This needs more
+      // debugging. The code below works and is idiomatic but could be prettier and more consistent
+      // with the use of TraceUtil-provided framework.
+      io.opentelemetry.api.trace.Span span =
```
[the remark this answers is not in the document]
Comment (jimit-j-shah) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:170]: - Looks like ReadWriteTransactionCallable is scheduled to run on an Executor in the `RetryHelper` library provided by gax layer. So that complicates the Context propagation.
- Also tried "manually" propagating the parent SpanContext to the "callable" but that doesn't work either as we don't have control over the future lifetimes (which run in the RetryHelper) so the parent span may `end()` before the child spans are executed(?).
- This needs some significant overhauling or gax support for context propagation - both are non trivial and wouldn't want to block the launch on this.

[the remark this answers is not in the document]
Comment (ehsannas) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:170]: Uh... not sure why using the TraceUtil directly doesn't work. Maybe it's because there's only 1 TraceUtil but several threads creating spans? If that's the case, can you try and see if this works:

Add an API `TraceUtil.isTracingEnabled()`, and then change:

```java
private TraceUtil.Span startSpanWithParentContext(
        String spanName,
        TraceUtil.SpanContext parentSpanContext) {
  TraceUtil otelTraceUtil = datastore.getOptions().getTraceUtil();
  if (!otelTraceUtil.isTracingEnabled()) {
    return new TraceUtil.Span();
  }

  SpanBuilder spanBuilder =
      otelTraceUtil
          .getTracer()
          .spanBuilder(com.google.cloud.datastore.telemetry.TraceUtil.SPAN_NAME_TRANSACTION_RUN)
          .setSpanKind(SpanKind.PRODUCER)
          .setParent(
              Context.current()
                  .with(
                      io.opentelemetry.api.trace.Span.wrap(
                          parentSpanContext.getSpanContext())));
  return TraceUtil.Span(otelTraceUtil.addSettingsAttributesToCurrentSpan(spanBuilder).startSpan());
}
```

```java
TraceUtil.Span span =  startSpanWithParentContext(
              com.google.cloud.datastore.telemetry.TraceUtil.SPAN_NAME_TRANSACTION_RUN,
              parentSpanContext);

try (TraceUtil.Scope ignored = span.makeCurrent()) {
...
```

This way the opentelemetry code will remain off the execution path if tracing is disabled, and OTEL types are also not used on the main code path.
